### ISSUE TITLE
design discussion: metadata index

### ISSUE THREAD

**Issue by houqp** [koreader/koreader#2122]
We have tried to keep things simple and avoid introducing indexes in KOReader. So far, it has been working well for us. But we might need to introduce it in the future especially with the new KOReader home. Let's use this issue to discuss various design tradeoffs for our metadata index module.

The first couple things that come to my mind are sqlite and leveldb/rocksdb.

**Comment by Hzj-jie** [comment]
Copy some of my comments from #1938

Briefly, I suggest to use a background task to build index file with language specific tools, instead of merging the logic in KOReader core logic.

So instead of adding a input method - as this by mentioned, instead of inputting Chinese, searching books with Chinese names is more important - I would suggest to use an index file. Then we can use background tasks to index file names or even file contents without changing KOReader core logic.

My initial plan is to use bash scripts, we can use various existing tools, grep, awk,cut, etc, without writting complex code.

**Comment by Hzj-jie** [comment]
So reply to HouQP's comment, I think for content indexing, which is even more complex, we can use sql lite. But I can image this is a long term change. But for title indexing, which I believe can fulfill 90% user requirements, grep and awk look like the right tools for a fast solution.

**Comment by houqp** [comment]
If we are going add indexes, which is a really big change, I would prefer we do it properly it from the beginning. We are an open-source project so there is no deadline and we can take time to engineer good design and code. Choosing a design because it's faster to implement will likely to cause tech dept down the road that will end up causing us more time to fix/refactor/migrate in the future.

Just my thoughts about using grep/awk:

* Maintainability: We either have to interact with them in Lua (hard to work with) or write shell scripts directly (hard to maintain and extend).
* Portability: keep in mind that we don't control userspace utils in all platforms. This means we will have to ship with our own grep and awk :(
* As you already mentioned, it won't cover all the use cases. This means we will end up with two index code paths if we want to add more advanced indexes in the future. It is not going to be fun to maintain this.

**Comment by houqp** [comment]
More idea: use index engine like lucene/Xapian. The official lucene is written in Java, so it relies on JVM. Xapian might be a better choice for us.

**Comment by Hzj-jie** [comment]
Emmm, good point. But I also do not think it's a good idea to add complex input method in current system. What do you think?

**Comment by houqp** [comment]
I totally agree we should make the input method a self contained module. After the next stable release, we should start working on splitting up KOReader so it is more resilient to crashes from sub-components.

**Comment by KenMaltby** [comment]
I hope that the fact that a number of Kobo owners prefer KOReader, in major part because it is not dependent upon a corruptible database.

**Comment by chrox** [comment]
I dislike the idea of whole-library full-text indexing neither.

Full-text indexing is problematic in embedded device with limited memory capacity and computation power and especially for application like Koreader that is run-time limited, the Koreader application itself could be terminated at anytime when it would be indexing (Kindle native system sometimes dies because of indexing even it runs the indexing task in a background process that is daemonized by the system, which cannot be implemented in Koreader, and the Kobo system more shitty than Kindle in this regard). I think this is not a problem with the persistent layer. [SQLite should perform perfectly well in single-user mostly-read workload](https://news.ycombinator.com/item?id=11934826). The problem is that indexing task is memory hungry, and consumer operating systems often have undefined behavior when there is not enough memory. And it won't be fixed by switching to another database.

My solution is that instead of indexing whole library we could only index documents that the users most probably use, documents in use or used in history. Broadly speaking, `history` data (or `metadata` after recent refactoring) are some kind of indexing that can be accessed directly by the reader modules. It does not really matter if we chose plain lua files or a SQLite for indexing result persistence. Supporting SQLite has another benefit that we can have a common facility to inter-operate with other datasources like the Vocabulary Builder in Kindle system which is also built on top of SQLite.

**Comment by chrox** [comment]
@Hzj-jie I still have no clue on how indexing could solve the Chinese input problem. Could you elaborate it more clearly?

**Comment by houqp** [comment]
That's what's in my mind too. Full document content index should be done out of band, i.e. done on the computer or from the cloud. Then ship those indexes to the reader as a read-only asset. That said, I think it should be fine to index file names and a tiny bit of metadata directly on the device.

People tend to think databases are easy to corrupt. IMO, that's a misconception. It is true that in many consumer devices, it's quiet common to run into a corrupted database. But that's because those softwares are not using it correctly.

One of the major design goals in database is to provide a durable storage mechanism so app developers don't have to worry about data corruption. In fact, our current way of storing book/reader settings through plain lua file is very easy to get into a corrupted state and loss data. If the process died in the middle of writing to disk, we ended up with a half written lua file and whatever data that hasn't been written are lost. We can of course improve this by first writing to a temporary file first then rename the temp file to replace the old one. That way it is at least more atomic than the current approach. But it still has many problems. For example, it is very inefficient because we have to write out everything every time. It also does not support role back or recovery in case of error.

A common design in database is to use write ahead log to persist transaction commands then apply those commands to update the datastore separately. This is way better than what we are doing right now. 

We can even improve the durability of this design by making the data update process almost atomic. The idea is to make everything read only and never change data in place. If everything is read only, then data update won't cause corruption. We can write all updates to a new place, then do an atomic pointer swap to point everything to the new data. The old data can be cleaned up later if there is no reference to it. This design not only increases the durability, but also makes it possible to revert back into any state in the past just like a version control system. The same design principal is used by closure, datomic and many other projects: https://www.youtube.com/watch?v=Cym4TZwTCNU. It is in fact quiet similar to leveldb's design.

Basically, what I want to say is if used/implemented correctly, database should be the ideal way to go. It will provide a more efficient and durable storage abstraction.

But talking is just talking, whoever write the actual code get to decide which engine to use, either an existing database or a custom one built from scratch :) I have no problem adopting sqlite if we keep the write transaction volume small and support corruption recovery in case of data loss. We don't need to create one giant sqlite db for everything. Each book and module settings can have their own db.

**Comment by Frenzie** [comment]
Closing in favor of #4951.
